### PR TITLE
New protocol design #2: Refactor of safe and whitelisting

### PR DESCRIPTION
**Pull request by bausano**

### CODE REVIEW

**Review by Suficio — APPROVED** [review]

File: sources/safe/royalties.move [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        amount: Balance<FT>,
+        /// The address where the amount should be transferred to.
+        /// This could be either the payment for the seller or a marketplace's
+        /// commision.
+        beneficiary: address,
+        /// Optionally we enable grouping of payments, e.g. if there are
+        /// multiple payments for one NFT (such as commission.), it might be
+        /// useful for the royalty collection logic to distinguish such
+        /// scenario.
+        trade: Option<ID>,
```
Comment (nmboavida) [inline sources/safe/royalties.move:44]: Where does the `ID` in`trade: Option<ID>` come from? Perhaps we can add some small explanantion on the comment

File: sources/safe/safe.move [hunk 2]
```diff
@@ -202,37 +202,19 @@ module nft_protocol::safe {
...
-
-        let (_, ref) = vec_map::remove(&mut safe.refs, &nft_id);
-        // an exlusively listed NFT must first have its transfer cap burned
-        // via fun burn_transfer_cap
-        assert_not_exlusively_listed(&ref);
-
-        let nft = object_bag::remove<ID, Nft<T, D>>(&mut safe.nfts, nft_id);
-        nft::transfer_to_owner(nft);
-    }
-
```
[the remark this answers is not in the document]
Comment (bausano) [reply sources/safe/safe.move:222]: We don't want to allow the user to freely take an NFT out of a safe at the current design stage, which is what this function was for.

File: sources/safe/royalties.move (hunk 1 above)
Comment (bausano) [reply sources/safe/royalties.move:44]: It could be any sort of data really, should be documented better, will add an issue

Comment (bausano) [reply sources/safe/royalties.move:44]: https://github.com/Origin-Byte/nft-protocol/issues/96

File: sources/safe/royalties.move [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    ///
+    /// Typically, this would be a witness exported from the collection contract
+    /// and it would access the balance to calculate the royalty in its custom
+    /// implementation.
+    public fun balance_mut<C, W: drop, FT>(
+        _witness: W,
+        payment: &mut TradePayment<C, W, FT>,
+    ): &mut Balance<FT> {
+        &mut payment.amount
+    }
```
[the remark this answers is not in the document]
Comment (bausano) [reply sources/safe/royalties.move:84]: You need access to `&mut balance` to be able to actually charge the royalty. Ie. you calculate it from this balance and then you take your cut out of it.

File: sources/nft/nft.move [hunk 4]
```diff
@@ -2,178 +2,78 @@
...
+    use nft_protocol::utils;
+    use std::type_name::{Self, TypeName};
+    use sui::bag::{Self, Bag};
+    use sui::object::{Self, UID};
+    use sui::transfer;
+    use sui::tx_context::{Self, TxContext};
 
-    // NFT object with an option to hold `D`ata object
-    struct Nft<phantom T, D: store> has key, store {
+    struct NFT<phantom C> has key, store {
```
[the remark this answers is not in the document]
Comment (bausano) [reply sources/nft/nft.move:15]: I believe an OTW must have `drop` only and no fields.

File: sources/utils/utils.move [hunk 5]
```diff
@@ -29,4 +31,40 @@ module nft_protocol::utils {
...
         new_vec
     }
+
+    /// One time witness is a type exported by a contract which follows the
+    /// module name.
+    ///
+    /// Witness is a type always in form "struct Witness has drop {}"
+    ///
+    /// They must be from the same module for this assertion to be ok.
+    public fun assert_same_package_as_witness<OneTimeWitness, Witness>() {
```
[the remark this answers is not in the document]
Comment (bausano) [reply sources/utils/utils.move]: renamed

**Pull request merged**